File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
class Queue {
    items = [];
    front = 0;
    rear = 0;
    
    push(item) {
        this.items.push(item);
        this.rear++;
    }
    
    pop() {
        return this.items[this.front++];
    }
    
    isEmpty() {
        return this.front === this.rear;
    }
}

function solution(maps) {
    const move = [[-1, 0], [0, -1], [0, 1], [1, 0]]; // 좌 하 우 상
    const n = maps.length; // 세로
    const m = maps[0].length; // 가로
    const dist = Array.from({ length: n }, () => Array(m).fill(-1));
    
    function bfs(start) {
        const q = new Queue();
        q.push(start);
        dist[start[0]][start[1]] = 1;
        
        while(!q.isEmpty()) {
            const here = q.pop();
            
            for(const [dx, dy] of move) {
                const row = here[0] + dx;
                const column = here[1] + dy;
                
                if(row < 0 || row >= n || column < 0 || column >= m) continue;
                if(maps[row][column] === 0) continue;
                
                if(dist[row][column] === -1) {
                    q.push([row, column]);
                    dist[row][column] = dist[here[0]][here[1]] + 1;
                }
            }
        }
    }
    
    bfs([0, 0]);
    
    console.log(dist)
    
    return dist[n - 1][m - 1];
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

move = [[1, 0], [0, -1], [-1, 0], [0, 1]]

def solution(maps):
    n, m = len(maps), len(maps[0])
    def is_valid_move(row, col):
        return 0 <= row < n and 0 <= col < m and maps[row][col] == 1
    visited = [[False] * m for _ in range(n)]
    q = deque() 
    q.append([0, 0, 1]) # [행, 열, 이동 횟수]
    visited[0][0] = True
    
    while q:
        y, x, count = q.popleft()
        if y == n - 1 and x == m - 1:
            return count
        
        for dy, dx in move:
            ny, nx = y + dy, x + dx
            if not is_valid_move(ny, nx):
                continue
                
            if not visited[ny][nx]:
                visited[ny][nx] = True
                q.append([ny, nx, count + 1])
    
    return -1